jobrunner/run.py

```diff
@@ -136,6 +186,97 @@ def handle_running_job(job):
             cleanup_job(job)
 
 
+def handle_running_job_api(job, api):
+    if job.cancelled:
+        log.info("Cancellation requested, killing job")
+        api.terminate(job_to_job_definition(job))
+
+    try:
+        is_running = sync_job_status(job, api)
+
+        if is_running:
+            set_message(job, "Running")
+        else:
+            mark_job_as_completed(job)
+            api.cleanup(job_to_job_definition(job))
+    except JobError as exception:
+        set_message(job, "Failed")
+        mark_job_as_failed(job, exception)
+        api.cleanup(job_to_job_definition(job))
+    except Exception:
+        set_message(job, "Failed")
+        mark_job_as_failed(job, "Internal error when finalising job")
+        # We don't clean up here, to facilitate with debugging this unexpected error
+        raise
+
+
+def job_to_job_definition(job):
+    action_args = shlex.split(job.run_command)
+    allow_database_access = False
+    env = {"OPENSAFELY_BACKEND": config.BACKEND}
+    # Check `is True` so we fail closed if we ever get anything else
+    if is_generate_cohort_command(action_args) is True:
+        if not config.USING_DUMMY_DATA_BACKEND:
+            allow_database_access = True
+            env["DATABASE_URL"] = config.DATABASE_URLS[job.database_name]
+            if config.TEMP_DATABASE_NAME:
+                env["TEMP_DATABASE_NAME"] = config.TEMP_DATABASE_NAME
+            if config.PRESTO_TLS_KEY and config.PRESTO_TLS_CERT:
+                env["PRESTO_TLS_CERT"] = config.PRESTO_TLS_CERT
+                env["PRESTO_TLS_KEY"] = config.PRESTO_TLS_KEY
+            if config.EMIS_ORGANISATION_HASH:
+                env["EMIS_ORGANISATION_HASH"] = config.EMIS_ORGANISATION_HASH
+    # Prepend registry name
+    image = action_args.pop(0)
+    full_image = f"{config.DOCKER_REGISTRY}/{image}"
+    if image.startswith("stata-mp"):
+        env["STATA_LICENSE"] = str(config.STATA_LICENSE)
+
+    # Jobs which are running reusable actions pull their code from the reusable
+    # action repo, all other jobs pull their code from the study repo
+    study = job_executor.Study(job.action_repo_url or job.repo_url, job.action_commit or job.commit)
+    # Both of action commit and repo_url should be set if either are
+    assert bool(job.action_commit) == bool(job.action_repo_url)
+
+    input_files = []
+    for action in job.requires_outputs_from:
+        for filename in list_outputs_from_action(action):
+            input_files.append(filename)
+
+    outputs = {}
+    for privacy_level, named_patterns in job.output_spec.items():
+        for name, pattern in named_patterns.items():
+            outputs[pattern] = privacy_level
+
+    return job_executor.JobDefinition(job.id, study, job.workspace,
```

Comment:
This is another change from the spike - I used `job.id` rather than `job.slug` for the JobDefinition.id